Redesign factory objects for batching

rmlarose opened the pull request:
Fixes #317.

rmlarose commented:
Supersedes #378 because I botched the git revert.

As discussed, coverage intentionally decreases until we fix the bad tests of Qiskit utils. Fixing Issue #382 will increase coverage again.

Ready for final review @andreamari @karalekas.

karalekas commented on docs/source/guide/guide-zne.rst:
```diff
@@ -287,13 +287,13 @@ The typical tasks of a factory are:
 3. Given the history of noise scale factors and results, evaluate the associated zero-noise extrapolation.
 
-The structure of the :class:`.Factory` class is adaptive by construction, since the choice of the next noise
+The structure of the :class:`.BaseFactory` class is adaptive by construction, since the choice of the next noise
```
```suggestion
The structure of the :class:`.Factory` class is adaptive by construction, since the choice of the next noise
```

karalekas commented on mitiq/zne/zne.py:
```diff
@@ -49,7 +49,7 @@ def execute_with_zne(
     if not isinstance(factory, Factory):
         raise TypeError(
-            f"Argument `factory` must be of type mitiq.factories.Factory "
+            f"Argument `factory` must be of type mitiq.factories.BaseFactory "
```
```suggestion
            f"Argument `factory` must be of type mitiq.factories.Factory "
```

karalekas reviewed (approved):
Two tiny holdovers. Didn't do a super thorough re-check of everything, but if it follows our discussion earlier it's good for me. The follow-up PR offers chances to fix things too if anything slipped through.

andreamari commented:
@rmlarose : I made some further changes see last commit:

- removed `next` and `is_converged` from `BatchedFactory` since I think they are not necessary
- Tried to simplify the structure by merging `run_batched` into the `run` method of `BatchedFactory`

What do you think? If I made something wrong, feel free to revert.

andreamari commented:
@rmlarose I got where is the bug. Will fix soon

andreamari commented:
This is now ready for re-review.
@rmlarose

rmlarose commented on mitiq/zne/inference.py:
```diff
@@ -420,11 +422,18 @@ def run_classical(
+        else:
+            self._outstack = [
+                scale_factor_to_expectation_value(scale_factor)
+                for i, scale_factor in enumerate(self._scale_factors)
```
`enumerate` isn't necessary

andreamari replied:
Thanks. This part is now removed.

rmlarose commented on mitiq/zne/inference.py:
```diff
@@ -454,38 +459,20 @@ def _batch_populate_instack(self) -> None:
-            )
-        return in_params
+        # Check if keywords are absent
+        if params is [{} for _ in params]:
+            return None
```
Why not return the list of empty dictionaries? This is perfectly fine and gets rid of a check for keyword args in line 395.

andreamari replied:
Thanks, good point.

rmlarose commented on mitiq/zne/inference.py:
```diff
@@ -157,6 +158,57 @@ def mitiq_polyfit(
+        results = []
+        for circ, kwargs in zip(batch, batched_kwargs):
+            results.append(executor(circ, **kwargs))
+        return results  # type: ignore
```
Also can be simplified by allowing `batched_kwargs` to be a list of empty dictionaries. See comment in `_get_keyword_args`.

rmlarose commented on mitiq/zne/tests/test_inference.py:
```diff
@@ -275,7 +275,7 @@ def test_run_sequential_and_batched(factory, batched):
     if batched:
 
-        def executor(circuits, kwargs) -> List[float]:
+        def executor(circuits, **kwargs) -> List[float]:
```
Since this is a batched executor, the argument `kwargs` is a list of keyword arguments, so the signature should be 

```python
def executor(circuits, kwargs) -> List[float]
```

Maybe `kwargs` can be renamed to make this clearer.

andreamari replied:
In the last commit I just removed the extra argument.
Note: I tried to write the code such that it is possible to define a batched executor without the `kwargs_list` argument, since this is "used" by a factory only in the particular case in which `shot_list` is used.

rmlarose reviewed (commented):
Thanks @andreamari. My main reservation is a conceptual one. Namely, `_make_batched_executor` does not actually make a batched executor - the second one returned is still a "single" or "sequential" executor because the circuits are run one-by-one. 

It may be useful to have such a function as a standalone, but a case for this isn't clear to me currently. For this reason I would argue it's more appropriate for this function, or whichever function handles the batching (`run_batched` before), to be a method of `BatchedFactory`.

andreamari commented:
Thanks @rmlarose!

- As suggested I removed the external `_make_executor_batched` function. I agree it could have been confusing.

- Now there is an internal `_is_executor_batched` method which is used in the body of the `run` method which does not mutate the executor but just detects its type.

- I changed a bit the `_get_keywords method` such that it directly takes into account the `num_to_average` repetitions. In this way, the output is directly compatible with the list of circuits returned by `_generate_circuits`.

- I tried to explain in the docs of `run` how to deal with keywords of a batched executor via a `kwargs_list` argument. Since this is actually relevant only if the `shot_list` option is used, I tried to allow the user to also define a betched executor without the  `kwargs_list` whenever this is possible. This point is a bit delicate and there could be a better/different solution.

Feel free to directly make changes if necessary.

rmlarose commented:
Thanks @andreamari! Had a quick look and I like these changes. I will now work on fixing the qiskit utils tests (#382) so we can merge this.

> * As suggested I removed the external `_make_executor_batched` function. I agree it could have been confusing.
> * Now there is an internal `_is_executor_batched` method which is used in the body of the `run` method which does not mutate the executor but just detects its type.

Great!

> * I changed a bit the `_get_keywords method` such that it directly takes into account the `num_to_average` repetitions. In this way, the output is directly compatible with the list of circuits returned by `_generate_circuits`.

This is much better imo, thanks for this.

> * I tried to explain in the docs of `run` how to deal with keywords of a batched executor via a `kwargs_list` argument. Since this is actually relevant only if the `shot_list` option is used, I tried to allow the user to also define a betched executor without the  `kwargs_list` whenever this is possible. This point is a bit delicate and there could be a better/different solution.

I don't have ideas yet but I'll think about this.

andreamari commented:
Note sure why tests are not running

rmlarose commented:
> Note sure why tests are not running

It may be due to the conflicts.

rmlarose commented:
@andreamari latest commit fixes the Qiskit utils tests. I'll try to resolve the conflicts now, which may be a bit involved due to the changes for strict mypy checking (#380).

rmlarose commented:
Ready for re-re-final-re-review @andreamari @karalekas

rmlarose commented on mitiq/mitiq_qiskit/tests/test_zne_mitiq_qiskit.py:
```diff
@@ -49,103 +41,113 @@ def measure(circuit, qid) -> QuantumCircuit:
     # Ensure that we have a classical register of enough size available
     if len(circuit.clbits) == 0:
-        reg = ClassicalRegister(qid + 1, "creg")
+        reg = ClassicalRegister(qid + 1, "cbits")
```
Note `creg` is a keyword in Qasm. This was causing bugs in the conversions.

andreamari replied:
nice catch!

rmlarose commented on mitiq/zne/inference.py:
```diff
@@ -202,6 +237,262 @@ def get_expectation_values(self) -> np.ndarray:
+        for d in params:
+            _ = d.pop("scale_factor")
+
+        # Repeat each keyward num_to_average times
```
```suggestion
        # Repeat each keyword num_to_average times
```

andreamari commented:
@karalekas, you approved this PR but then we made other changes. Could you do a final check on this?

karalekas commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@
 
 ## Version 0.3.0 (In Development)
 
+- Redesign factories, batch runs in BatchedFactory, fix Qiskit utils tests (@rmlarose, @andreamari, gh-381).
```
this is in the wrong spot -- will need to resolve merge conflicts when moving to "All Changes" section

karalekas commented on mitiq/zne/__init__.py:
```diff
@@ -14,3 +14,4 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 from mitiq.zne.zne import execute_with_zne, mitigate_executor, zne_decorator
+from mitiq.zne import scaling
```
i dont understand what this is for

karalekas commented on mitiq/zne/inference.py:
```diff
@@ -202,6 +237,262 @@ def get_expectation_values(self) -> np.ndarray:
+        return self
+
+    def run_classical(
+        self, scale_factor_to_expectation_value: Callable[..., float]
```
is this `scale_factor_to_expectation_value` the same as `noise_to_expval` in `AdaptiveFactory`? if so, (1) why do they have different names and (2) why isnt this method defined on `Factory` as an `@abstractmethod`

rmlarose replied:
Renamed to `scale_factor_to_expectation_value` and defined the abstract method.

karalekas commented on mitiq/zne/inference.py:
```diff
@@ -202,6 +237,262 @@ def get_expectation_values(self) -> np.ndarray:
+
+        return self
+
+    def run_classical(
```
renaming this from `iterate` is a backwards-incompatible change. I would recommend adding back an `iterate` method to `Factory` that (1) raises a `DeprecationWarning` saying that `iterate` will be removed in `v0.4.0`, and (2) then calls `run_classical`

rmlarose replied:
Deprecated iterate.

karalekas reviewed (changes requested):
A couple small things.

karalekas commented on requirements.txt:
```diff
@@ -6,6 +6,9 @@ cirq-unstable==0.9.0.dev20200813185757
 pyquil~=2.18.0
 qiskit~=0.16.2
 
+# to handle deprecations
+deprecation~=2.1.0
```
this is very cool, but it will need to be in `install_requires` of the `setup.py` (otherwise `mitiq` will break), and I'm not sure that it really makes life so much easier that it's worth the additional dependency

rmlarose replied:
What's the alternative for handling deprecation that doesn't require an additional dependency?

karalekas replied:
Typically one would just use the `warnings` library to raise a `DeprecationWarning` like we do in pyQuil [here](https://github.com/rigetti/pyquil/blob/6f5c9b1a3aced2b5cec5d81859d81c0e339d16fb/pyquil/device/_main.py#L106). The `deprecations` decorator likely does exactly this, but with a nicely formatted string with the version numbers indicating when to expect the feature to be removed (which you can do yourself with the string argument to `warn`). Lmk if you have any more qs!

karalekas reviewed (changes requested):
Also, what's up with coverage? Can we make it pass?

rmlarose commented:
> Also, what's up with coverage? Can we make it pass?

Not sure why codecov is failing now @karalekas. The comment says coverage increased. 

> Merging #381 into master will increase coverage by 1.32%.
> The diff coverage is 93.25%.

karalekas reviewed (approved):
For some reason it wants the diff to have 93.55% coverage instead of the 93.25% coverage that it currently has. I'm fine with letting that go in the interest of moving forward. LGTM!

nathanshammah commented:
Catching up with this now. This is great @rmlarose, and great review and additions from @andreamari and @karalekas.
